[Concept] Docs - Classes (#649)
exercism/cpp: concepts/classes/introduction.md

# Introduction

It is time to get to one of the core principles of C++: object-oriented programming.
OOP is centered around classes - user-defined types of data with their own set of functions.
We will start with the basics and will cover more advanced topics further down the syllabus tree.

## Members

Classes can have **member variables** and **member functions**.
They are accessed by the **member selection** operator `.`.
Just as variables outside of classes, it is advisable to initialize them with a value upon declaration.
This value will then become the default for newly created objects from this class.

## Encapsulation and Information Hiding

Classes offer the option to restrict the usage of their members.

> The two basic cases are `private` and `public`.

Author: I was hoping the example makes it clear and I don't (yet) have to go into the nitpicking that `protected` brings, when it is introduced. You are probably right it should get a bit more explanation

> The two basic cases are `private` and `public`.

Reviewer: It might just be my penchant for over-explanation surfacing.  😉 

I do think saving `protected` for when you have a very clear use case would be good.  I feel like `public`/`private`/`protected` often get swirled around and confused when someone is learning them all at once.

> The two basic cases are `private` and `public`.

Author: Better?

> The two basic cases are `private` and `public`.

Reviewer: Yes!  I like the change.  😄

Everything is private by default and needs to be explicitly marked to be usable outside of the class.

## Basic example

The definition of a new `class` can be seen in the following example.
Notice the `;` after the definition.

```cpp
class Wizzard {
  public:
    int cast_spell() {
      return damage;
    }
    std::string name{};
  private:
    int damage{5};
};

```

You can access all variables inside the class.
Look at `damage` in the `cast_spell` function.
You cannot read or change `private`  members outside of the class:

```cpp
Wizzard silverhand{};
// calling the `cast_spell` function is okay, it is public:
iggwilv.cast_spell();
// => 5
// name is public and can be changed:
silverhand.name = "Laeral";
// damage is private -> Compilation error
silverhand.damage = 500; 
```

## Constructors

Constructors offer the possibility to assign values to member variables when the object is created.
They have the same name as the class and no return type.

> A class can have several constructors:

Author: Is this enough?

```cpp
class Wizzard {
  public:
    Wizzard(std::string new_name) {
      name = new_name;
    }
    Wizzard(std::string new_name, int new_damage) {
      name = new_name;
      damage = new_damage;
    }
    int cast_spell() {
      return damage;
    }
    std::string name{};
  private:
    int damage{5};
};

Wizzard el{"Eleven"};       // deals  5 damage
Wizzard vecna{"Vecna", 50}; // deals 50 damage
```

Constructors are a big topic and have many nuances.
If you are not explicitly defining a constructor for your class, the compiler will do the job for you.

> This has happened in the first example above.

Author: Do you think that will bring additional insight? Because the compiler would insert:
```cpp
Wizard(){};
```

> This has happened in the first example above.

Reviewer: ah!  And that is the misunderstanding here.  When I first read this, I took it to mean that if you didn't provide:

```c++
    Wizzard(std::string new_name) {
      name = new_name;
    }
```
..the compiler would fill in for you.  But the compiler will only provide an _empty_ constructor, hence the mention of setting member variable values.

> This has happened in the first example above.

Author: What do you think of this version?

> This has happened in the first example above.

Reviewer: Yup. That works.  And when all else fails, you can comment.  But I do think this version reads better.

All variables are set to the value that was stated in the definition of the class.
If you had not given any values, the variables would be uninitialized, which might have unintended consequences.

## Structs

> Structs are a remainder of the C past that developed into C++.

Reviewer: ```suggestion
Structs are a holdover from the C past that developed into C++.
```

> They are pretty much the same thing as classes, there is only one exception.

Reviewer: ```suggestion
They are effectively the same thing as `classes` with one important exception.
```

> By default everything in a class is private. Structs are public until defined otherwise.

Reviewer: ```suggestion
By default, everything in a `class` is `private`.
Structs, on the other hand are `public` until defined otherwise.
```

> Conventionally, the struct keyword is often used for data-only structures.

Reviewer: ```suggestion
Conventionally, the `struct` keyword is often used for **data-only structures**.
```

> The class keyword is preferred for objects that have both data and functions.

Reviewer: ```suggestion
The `class` keyword is preferred for objects that have both data **and** function members.
```

# Introduction

> It is time to get to one of the core principles of C++: object-oriented programming (OOP).

Reviewer 2: "principles" -> "paradigms" (at least that's what it's most often called in my experience)

OOP is centered around `classes` - user-defined types of data with their own set of related functions.
We will start with the basics and will cover more advanced topics further down the syllabus tree.

## Members

Classes can have **member variables** and **member functions**.
They are accessed by the **member selection** operator `.`.
Just as variables outside of `classes`, it is advisable to initialize member variables with a value upon declaration.
This value will then become the default for newly created objects from this class.

## Encapsulation and Information Hiding

Classes offer the option to restrict access to their members.
The two basic cases are `private` and `public`.
All members are `private` by default and need to be explicitly marked to be usable outside of the class.

## Basic example

The definition of a new `class` can be seen in the following example.
Notice the `;` after the definition.

```cpp
class Wizzard {
  public:
    int cast_spell() {
      return damage;
    }
    std::string name{};
  private:
    int damage{5};
};

> 

Reviewer 2: You could consider adding comments to this code example, explaining key lines.
E.g. something like
```
  public:   // from here on all members are accessible publicly
    inc cast_spell() {  // defines the public member function cast_spell
    ...
    std::string name{};  // defines the public member variable `name`
  private:  // from here on all members are private
    int damage{5};  // defines the private member variable `damage`
```

Author: Good idea! This would help indeed

```cpp
Wizzard silverhand{};
// calling the `cast_spell` function is okay, it is public:
iggwilv.cast_spell();
// => 5
// name is public and can be changed:
silverhand.name = "Laeral";
// damage is private -> Compilation error
silverhand.damage = 500; 
```

## Constructors

Constructors offer the possibility to assign values to member variables at object creation.
They have the same name as the `class` and do not have return type.
A class can have several constructors:

```cpp
class Wizzard {
  public:
    Wizzard(std::string new_name) {
      name = new_name;
    }
    Wizzard(std::string new_name, int new_damage) {
      name = new_name;
      damage = new_damage;
    }
    int cast_spell() {
      return damage;
    }
    std::string name{};
  private:
    int damage{5};
};

Wizzard el{"Eleven"};       // deals  5 damage
Wizzard vecna{"Vecna", 50}; // deals 50 damage
```

Constructors are a big topic and have many nuances.
If you are not explicitly defining a `constructor` for your `class`, the compiler will do the job for you.
This has happened in the first example above.
All variables are set to the value that was stated in the definition of the class.

> If you had not given any values in that definition, the variables would be uninitialized, which might have unintended consequences.

Reviewer 2: "would" -> "might" because all types with a default constructor will be default-initialized

## Structs

> Structs are a holdover from the C past that developed into C++.

Reviewer 2: that sounds like structs are somehow discouraged. but they are even recommended by the C++ Core Guidelines for compound entities that do not have an invariant (i.e. when they just hold a couple of member variables that can be modified freely)

Author: ```suggestion
Structs came from the language's original C roots and are as old as C++ itself.
```
How does that sound to you?

## Encapsulation and Information Hiding

Classes offer the option to restrict access to their members.
The two basic cases are `private` and `public`.
`private` members are not accessible from outside the class.
`public` members can be called freely.
All members are `private` by default and need to be explicitly marked to be usable outside of the class.

## Basic example

The definition of a new `class` can be seen in the following example.
Notice the `;` after the definition.

```cpp
class Wizzard {
  public:
    int cast_spell() {
      return damage;
    }
    std::string name{};
  private:
    int damage{5};
};

```

You can access all variables from within the class.
Take a look at `damage` inside the `cast_spell` function.
You cannot read or change `private`  members outside of the class:

```cpp
Wizzard silverhand{};
// calling the `cast_spell` function is okay, it is public:
iggwilv.cast_spell();
// => 5
// name is public and can be changed:
silverhand.name = "Laeral";
// damage is private -> Compilation error
silverhand.damage = 500; 
```

## Constructors

Constructors offer the possibility to assign values to member variables at object creation.

> They have the same name as the `class` and do not have return type.

Reviewer: ```suggestion
They have the same name as the `class` and do not have a return type.
```

A class can have several constructors.

> This is useful if you are not always in need to set all variables.

Reviewer: ```suggestion
This is useful if you do not always have a need to set all variables.
```

    }
    int cast_spell() {
      return damage;
    }
    std::string name{};
  private:
    int damage{5};
};

Wizzard el{"Eleven"};       // deals  5 damage
Wizzard vecna{"Vecna", 50}; // deals 50 damage
```

Constructors are a big topic and have many nuances.
If you are not explicitly defining a `constructor` for your `class`, then - and only then - the compiler will do the job for you.
This has happened in the first example above.
The _silverhand_ object is created by calling the default constructor, no arguments were passed.
All variables are set to the value that was stated in the definition of the class.
If you had not given any values in that definition, the variables would be uninitialized, which might have unintended consequences.

## Structs

Structs are a holdover from the C past that developed into C++.
They are effectively the same thing as `classes` with one important exception.
By default, everything in a `class` is `private`.
Structs, on the other hand are `public` until defined otherwise.
Conventionally, the `struct` keyword is often used for **data-only structures**.
The `class` keyword is preferred for objects that have both data **and** function members.

> The class keyword is preferred for objects that have both data and functions.

Reviewer: I think either line 98 or line 99 .. but not both.

Reviewer 2: @vaeng Oh, I completely missed these lines.  
Are you sure about them? In my experience it's completely fine for a `struct` to have member variables and member functions.

From a technical standpoint a `struct` and a `class` are almost identical, the only differences are that the implicit "default" access specifier (`public` for `struct`, `private` for `class`), and an implicit "default" access-specifier when defining a derive class (`public` inheritance when deriving from a `struct`, `private` inheritance when deriving from a `class`).

So it's "just" about conventions. The advice I heard from leading experts is that classes are for types with an invariant, something that must be true over the lifetime of an object. The constructor is responsible for establishing that invariant and it must hold at the end of each public member function. That's when you need to make member variable `private`, it allows the class to control the values and preserve the invariant. structs are for types without an invariant, where member variables can be modified from the outside. But that doesn't necessarily prevent them for having member functions.

Author: You are right. My "convention" is just the way I practiced it in the past.
The "use structs for pod" is surely a consequence of something without an invariant. I wonder how we can write that advice without being too technical.
I went through a lot of math courses and know people who did them with me who never became confident in finding invariants for their logical reasoning. 
Maybe I should give an example like "a wizard's life or mana cannot be negative, so these values are controlled by functions, rather than offer the possibility to set those directly". That might be a bit too wordy? Should this be moved the about.md?

Author: What do you think of this change? @siebenschlaefer @BethanyG

Reviewer 2: Looks good!
